internal/client/state_store.go

```diff
@@ -0,0 +1,93 @@
+package client
+
+import (
+	"errors"
+
+	"maunium.net/go/mautrix"
+	"maunium.net/go/mautrix/event"
+	"maunium.net/go/mautrix/id"
+)
+
+// StateStore implements the StateStore interface for OlmMachine.
+// It is used to determine which rooms are encrypted and which rooms are shared with a user.
+// The state is updated by /sync responses.
+type StateStore struct {
+	Storer *mautrix.InMemoryStore
+}
+
+// GetEncryptionEvent returns the encryption event for a room.
+func (ss *StateStore) GetEncryptionEvent(roomID id.RoomID) *event.EncryptionEventContent {
+	room := ss.Storer.LoadRoom(roomID)
+	if room == nil {
+		return nil
+	}
+	if evts, ok := room.State[event.StateEncryption]; ok {
+		if evt, ok := evts[""]; ok {
+			return evt.Content.AsEncryption()
+		}
+	}
+	return nil
+}
+
+// IsEncrypted returns whether a room has been encrypted.
+func (ss *StateStore) IsEncrypted(roomID id.RoomID) bool {
+	room := ss.Storer.LoadRoom(roomID)
+	if room == nil {
+		return false
+	}
+	_, ok := room.State[event.StateEncryption]
+	return ok
+}
+
+// FindSharedRooms returns a list of room IDs that the given user ID is also a member of.
+func (ss *StateStore) FindSharedRooms(userID id.UserID) []id.RoomID {
+	sharedRooms := make([]id.RoomID, 0)
+	for roomID, room := range ss.Storer.Rooms {
+		if room.GetMembershipState(userID) != event.MembershipLeave {
```

Comment:
What about invite? That isn't shared surely. Same for Ban. Only Join should apply here.